### ISSUE TITLE
Research implementing on-demand cross-platform e2e Cypress test execution on local dev environment

### ISSUE THREAD

**Issue by pjhill** [department-of-veterans-affairs/va.gov-team#19380]
# Description 
Given that VSP has access to BrowserStack via their open source repository license, use the [write-up provided by BrowserStack](https://www.browserstack.com/docs/automate/cypress) to install, configure, and document the BrowserStack Cypress CLI.

# Acceptance Criteria
- [x] VSP's approach to installing, configuring, and using the BrowserStack Cypress CLI package is documented in a markdown file in va.gov-team
- [ ] Test execution of Cypress E2E test suite in all recommended most used desktop environments through the BrowserStack Cypress CLI

**Comment by holdenhinkle** [comment]
Branches:
- `va.gov-team` (documentation) - https://github.com/department-of-veterans-affairs/va.gov-team/compare/19380_documentation_for_browserstack
- `vets-website` - https://github.com/department-of-veterans-affairs/vets-website/compare/19380_run_cyress_tests_on_browserstack

Google doc:
- Google Analytics VA.gov Platform (Browser/OS) Data - https://docs.google.com/spreadsheets/d/1v77a8BBAwcNCQcJyUCg1tzyzwDngAUQt3G72ELcT1Qg/edit?usp=sharing

**Comment by holdenhinkle** [comment]
I submitted a support ticket with BrowserStack regarding the following error over 24 hours ago:

```
The support file is missing or invalid.

Your `supportFile` is set to `.\src\platform\testing\e2e\cypress\support\index.js`, but either the file is missing or it's invalid. The `supportFile` must be a `.js`, `.ts`, `.coffee` file or be supported by your preprocessor plugin (if configured).

Correct your `cypress.json`, create the appropriate file, or set `supportFile` to `false` if a support file is not necessary for your project.

Or you might have renamed the extension of your `supportFile` to `.ts`. If that's the case, restart the test runner.

Learn more at https://on.cypress.io/support-file-missing-or-invalid
```

**Comment by holdenhinkle** [comment]
I received the following email from BrowerStack this morning:

6:34 AM (5 hours ago)
to me

Hi Holden,
​
Yes, we're referring to browser combinations when we mention tests/parallels. 
​
With regard to the support/index.js error, we are checking this and will share our findings/updates soon. 

Regards,
Reehan
BrowserStack Support

**Comment by holdenhinkle** [comment]
List of OSes available for BrowserStack OS / Browser testing - https://www.browserstack.com/list-of-browsers-and-platforms/cypress_testing

List of browsers we should support with BrowserStack:

```
        {
            "browser": "chrome",
            "os": "Windows 10",
            "versions": [
                "latest",
                "latest-1"
            ]
        },
        {
            "browser": "edge",
            "os": "Windows 10",
            "versions": [
                "latest",
                "latest-1"
            ]
        },
        {
            "browser": "firefox",
            "os": "Windows 10",
            "versions": [
                "latest",
                "latest-1"
            ]
        },
        {
            "browser": "chrome",
            "os": "OS X Catalina",
            "versions": [
                "latest",
                "latest-1"
            ]
        }
```

**Comment by holdenhinkle** [comment]
I did a little more digging this morning and found this - https://github.com/browserstack/browserstack-cypress-cli/issues/76#issuecomment-709697862

We can’t add `'./commands'to the browserstack.json` file which imports these custom commands:

```
import './axeCheck';
import './expandAccordions';
import './login';
import './syncFixtures';
import './upload';
import './injectAxeThenAxeCheck';
import './viewportPreset';
import './hasCount';
import 'cy-mobile-commands';
```

I just sent the following messages to BrowserStack:

### Email 1
Hi there,

Here are the dependencies that we're pulling in, in our support/index.js file:

import '@testing-library/cypress/add-commands';
import 'cypress-axe';
import 'cypress-plugin-tab';

import './commands';

import './commands' imports all of our cypress custom commands.

I found this issue which might explain why we're getting the error - https://github.com/browserstack/browserstack-cypress-cli/issues/76#issuecomment-709697862

Can you please confirm that 1) this issue is probably why we're getting the error, 2) that this issue still exists and that you haven't addressed it? If you have addressed it, please let me know how to import the private module.

Thank you very much,

Holden

### Email 2
One other thing, regarding @testing-library/cypress module:

Here's our npm_dependencies block in browserstack.json: 

    "npm_dependencies": {
      "@testing-library/cypress": "^7.0.1",
      "cypress-axe": "^0.12.0",
      "cypress-plugin-tab": "^1.0.5"
    },

Then, in our support/index.js file, we import it like this, according to the documentation - https://www.npmjs.com/package/@testing-library/cypress:

import '@testing-library/cypress/add-commands';

I think this might be another reason we're getting the error message. Is there a work around?

Thanks,

Holden

### Thoughts
I suspect there are multiple reasons we're getting the error.

Instead of importing all the Cypress custom commands, I added them directly to `support/index.js` but that didn't fix the error. Also, the `login` custom command has an import (`import { VA_FORM_IDS } from 'platform/forms/constants';`), so that adds another monkey wrench into things.

Unfortunately, things aren't looking good right now...

**Comment by pjhill** [comment]
Would converting Cypress customizations to a public NPM package work around this issue?

**Comment by holdenhinkle** [comment]
Peter H, Darius D, and I had a meeting on Friday to discuss the BrowserStack/Cypress blockers to see if we could think of a way to address them.

The plan:
I'm going to comment out the imports that BrowserStack won't let us import into `support/index.js` (Cypress Testing Library commands and our custom commands). I’m going find a simple test that doesn’t use any Cypress Testing Library commands or custom commands and see if I can get that to run.

If I can get it to run, then I’ll find a test that uses Cypress custom commands and import them manually, and see if I can get that to run.

If that works, and we decide we're happy with this solution, Cypress tests that use Cypress custom commands will have to be quickly refactored -- the custom commands that are used for each test will have to be manually imported. Also, we'll have to create our own NPM package that has just has the Cypress Testing Library commands (not the entire Cypress Testing Library).

**Comment by holdenhinkle** [comment]
I tried the most simple test which produces a ‘pass’ in the headless test runner:

```
describe('COVID-19 Vaccination Preparation Form', () => {
  it('should fail to unsubscribe the user', () => {
  });
});
```

```
       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  applications/coronavirus-vaccinatio       41ms        1        1        -        -        - │
  │    n/tests/e2e/unsubscribeFail.cypress                                                         │
  │    .spec.js                                                                                    │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                         41ms        1        1        -        -        -  
✨  Done in 12.30s.
But it doesn’t run on BrowserStack:
Failed / skipped test report:
+--------------------------------------------------------------+--------+---------+-------------------------+
| Spec                                                         | Status | Browser | BrowserStack Session ID |
+--------------------------------------------------------------+--------+---------+-------------------------+
Total tests: 0, passed: 0, failed: 0, skipped: 0
Done in 123.77 seconds using 8 machines
Error:
The support file is missing or invalid.
Your `supportFile` is set to `.\src\platform\testing\e2e\cypress\support\index.js`, but either the file is missing or it's invalid. The `supportFile` must be a `.js`, `.ts`, `.coffee` file or be supported by your preprocessor plugin (if configured).
Correct your `cypress.json`, create the appropriate file, or set `supportFile` to `false` if a support file is not necessary for your project.
Or you might have renamed the extension of your `supportFile` to `.ts`. If that's the case, restart the test runner.
Learn more at https://on.cypress.io/support-file-missing-or-invalid
:slightly_frowning_face:

**Comment by holdenhinkle** [comment]
The [BrowserStack / Cypress Kitchen sink tutorial](https://www.browserstack.com/docs/automate/cypress/sample-tutorial) uses the following `support/index.js` file which imports Cypress custom commands, like the `vets-website` `support/index.js` file:

```
// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

// Import commands.js using ES2015 syntax:
import './commands'

// Alternatively you can use CommonJS syntax:
// require('./commands')
```

Here's the `browserstack.json` file `run-settings`:
```
  "run_settings": {
    "cypress_config_file": "./cypress.json",
    "cypress_version": "6",
    "project_name": "Cypress Kitchen Sink Example",
    "build_name": "Build no: 1",
    "parallels": 5,
    "npm_dependencies": {}
  },
```

It appears that importing ` './commands'` shouldn't be a problem after all.

-----

For the heck of it, I added all npm modules found in the `vets-website` `package.json` file to the `browserstack.json` `"npm_dependencies"` object and got a new error:

```npm WARN deprecated babel-eslint@10.1.0: babel-eslint is now @babel/eslint-parser. This package will no longer receive updates.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated metalsmith-permalinks@0.5.0: No longer maintained; Migrate to the following package: @metalsmith/permalinks
npm WARN deprecated url-search-params@1.1.0: now available as @ungap/url-search-params
npm WARN deprecated cheerio-select-tmp@0.1.1: Use cheerio-select instead
npm WARN deprecated graceful-fs@2.0.3: please upgrade to graceful-fs 4 for compatibility with current and future versions of Node.js
npm WARN deprecated axios@0.20.0: Critical security vulnerability fixed in v0.21.1. For more information, see https://github.com/axios/axios/pull/3410
npm WARN deprecated ecstatic@1.4.1: https://github.com/jfhbrook/node-ecstatic/issues/259
npm WARN deprecated request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated formatio@1.2.0: This package is unmaintained. Use @sinonjs/formatio instead
npm WARN deprecated samsam@1.3.0: This package has been deprecated in favour of @sinonjs/samsam
npm WARN deprecated text-encoding@0.6.4: no longer maintained
npm WARN deprecated chokidar@2.1.8: Chokidar 2 will break on node v14+. Upgrade to chokidar 3 with 15x less dependencies.
npm WARN deprecated node-pre-gyp@0.17.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
npm WARN deprecated request@2.88.0: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated mkdirp@0.5.1: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated jade@0.26.3: Jade has been renamed to pug, please install the latest version of pug instead of jade
npm WARN deprecated to-iso-string@0.0.2: to-iso-string has been deprecated, use @segment/to-iso-string instead.
npm WARN deprecated fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated json3@3.3.2: Please use the native JSON object instead of JSON 3
npm WARN deprecated wrench@1.3.9: wrench.js is deprecated! You should check out fs-extra (https://github.com/jprichardson/node-fs-extra) for any operations you were using wrench for. Thanks for all the usage over the years.
npm WARN deprecated fsevents@1.2.13: fsevents 1 will break on node v14+ and could be using insecure binaries. Upgrade to fsevents 2.
npm WARN deprecated coffee-script@1.12.7: CoffeeScript on NPM has moved to "coffeescript" (no hyphen)
npm WARN deprecated core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
npm WARN deprecated core-js@1.2.7: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
npm WARN deprecated mkdirp@0.3.0: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated kleur@2.0.2: Please upgrade to kleur@3 or migrate to 'ansi-colors' if you prefer the old syntax. Visit <https://github.com/lukeed/kleur/releases/tag/v3.0.0\> for migration path(s).
npm WARN deprecated socks@1.1.10: If using 2.x branch, please upgrade to at least 2.1.6 to avoid a serious bug with socket data flow and an import issue introduced in 2.1.0
npm WARN deprecated left-pad@1.3.0: use String.prototype.padStart()
npm WARN deprecated circular-json@0.3.3: CircularJSON is in maintenance only, flatted is its successor.
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated cross-spawn-async@2.2.5: cross-spawn no longer requires a build toolchain, use it instead
npm ERR! code EPERM
npm ERR! syscall spawn
npm ERR! errno EPERM
npm ERR! Error: spawn EPERM
npm ERR!     at ChildProcess.spawn (internal/child_process.js:373:11)
npm ERR!     at Object.spawn (child_process.js:549:9)
npm ERR!     at \cypress_node\node_modules\npm\node_modules\pacote\lib\util\git.js:216:24
npm ERR!     at \cypress_node\node_modules\npm\node_modules\promise-retry\index.js:29:24
npm ERR!  Error: spawn EPERM
npm ERR!     at ChildProcess.spawn (internal/child_process.js:373:11)
npm ERR!     at Object.spawn (child_process.js:549:9)
npm ERR!     at \cypress_node\node_modules\npm\node_modules\pacote\lib\util\git.js:216:24
npm ERR!     at \cypress_node\node_modules\npm\node_modules\promise-retry\index.js:29:24 {
npm ERR!   stack: 'Error: spawn EPERM\n    at ChildProcess.spawn ' +
npm ERR!     '(internal/child_process.js:373:11)\n    at Object.spawn ' +
npm ERR!     '(child_process.js:549:9)\n    at ' +
npm ERR!     '\\Windows\\cypress_node\\node_modules\\npm\\node_modules\\pacote\\lib\\util\\git.js:216:24\n' +
npm ERR!     '    at ' +
npm ERR!     '\\Windows\\cypress_node\\node_modules\\npm\\node_modules\\promise-retry\\index.js:29:24',
npm ERR!   errno: 'EPERM',
npm ERR!   code: 'EPERM',
npm ERR!   syscall: 'spawn'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It's possible that the file was already in use (by a text editor or antivirus),
npm ERR! or that you lack permissions to access it.
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.
```

Note that the [BrowserStack / Cypress Kitchen sink tutorial](https://www.browserstack.com/docs/automate/cypress/sample-tutorial) has a number of npm packages listed in the `package.json` file but none are included in the `browserstack.json` `"npm_dependencies"` object.

-----

I added `"supportFile": false,` to the `run-settings` object. That didn't work.

-----

BrowserStack finally replied to my last email (over a week later):

6:19 AM (7 hours ago)
to me

Hi Holden,
​
Thank you for the details.
​
I would like to share if while executing the test locally if you have mentioned something in package.json and the test is working, the same should work on BrowserStack as well after you specify the details in browserstack.json.
​
However, in order to debug this further, could you kindly share the corresponding build ID?
​
Awaiting your response.

Regards,
Heena Nanwani
BrowserStack Support

-----

I'm at a loss as to how to move forward.

**Comment by holdenhinkle** [comment]
Darius and I just paired on BrowserStack and we’re confident that we’ve exhausted all obvious options trying to get it to work. Creating an npm package wouldn’t do anything for us right now because we don’t know why it’s not working.

Two big things:
1. It doesn't like our `support/index.js` file, even when we comment the contents of it out.
2. If we set `supportFile` to `false` we stop getting the vague error about the `support/index.js` file not existing or being invalid and then we start getting errors about it not being able to find our tests, even when we try to test a specific spec:

```
Can't run because no spec files were found.

We searched for any files matching this glob pattern:

src\applications\coronavirus-vaccination\tests\e2e\unsubscribeFail.cypress.spec.js
```

```
Can't run because no spec files were found.

We searched for any files matching this glob pattern:

**\tests\**\*.cypress.spec.js
```

```
Can't run because no spec files were found.

We searched for any files matching this glob pattern:

**\*
```
etc.

Specify spec files to run - https://www.browserstack.com/docs/automate/cypress/specs-to-run
